metricproviders/kayenta/kayenta.go

```diff
@@ -0,0 +1,253 @@
+		if err != nil {
+			return metricutil.MarkMeasurementError(newMeasurement, err)
+		}
+		scopes = scopes + fmt.Sprintf(scopeFormat, name, string(controlScope), string(experimentScope))
+		if i < (len(metric.Provider.Kayenta.Scopes) - 1) {
+			scopes = scopes + ","
+		}
+
+	}
+
+	jobPayLoad := fmt.Sprintf(jobPayloadFormat, scopes, metric.Provider.Kayenta.Threshold.Pass, metric.Provider.Kayenta.Threshold.Marginal)
+
+	jsonValue, err := templateutil.ResolveArgs(jobPayLoad, run.Spec.Args)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	response, err := p.client.Post(jobURL, "application/json", bytes.NewBuffer([]byte(jsonValue)))
+	if err != nil || response.Body == nil || response.StatusCode != 200 {
+		if err == nil {
+			err = errors.New("Invalid Response")
+		}
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+	var dat map[string]interface{}
+	if err := json.Unmarshal(data, &dat); err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+	jobId := dat["canaryExecutionId"]
+	m := make(map[string]string)
+	m["canaryExecutionId"] = fmt.Sprintf("%v", jobId)
+	if len(m["canaryExecutionId"]) == 0 {
+
+		return metricutil.MarkMeasurementError(newMeasurement, errors.New("Invalid canaryExecutionId"))
+	}
+	newMeasurement.Metadata = m
+
+	newMeasurement.Phase = v1alpha1.AnalysisPhaseRunning
+
+	resumeTime := metav1.NewTime(time.Now().Add(resumeDelay))
+	newMeasurement.ResumeAt = &resumeTime
+	finishTime := metav1.Now()
+	newMeasurement.FinishedAt = &finishTime
+
+	return newMeasurement
+}
+
+// Resume should not be used the kayenta provider since all the work should occur in the Run method
+func (p *Provider) Resume(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, measurement v1alpha1.Measurement) v1alpha1.Measurement {
+
+	scoreURL := fmt.Sprintf(scoreURLFormat, metric.Provider.Kayenta.Address, measurement.Metadata["canaryExecutionId"])
+
+	response, err := p.client.Get(scoreURL)
+	if err != nil || response.Body == nil || response.StatusCode != 200 {
+		if err == nil {
+			err = errors.New("Invalid Response")
+		}
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
+
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
+
+	patch := make(map[string]interface{})
+
+	err = json.Unmarshal(data, &patch)
+	if err != nil {
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
+
+	score, ok, err := unstructured.NestedFloat64(patch, "result", "judgeResult", "score", "score")
+
+	if ok {
+		score := int(score)
+		measurement.Value = fmt.Sprintf("%v", score)
+		measurement.Phase = evaluateResult(score, metric.Provider.Kayenta.Threshold.Pass, metric.Provider.Kayenta.Threshold.Marginal)
+	} else {
+		if err == nil {
+			err = errors.New("Missing Score")
```

Comment:
I just realized that we didn't code towards the case where the Canary Execution is not finished when the Analysis controller runs Resume. If the canary run is not finished, then we should set the ResumeAt to 15 seconds in the future and leave the measurement unfinished instead of failing. The controller will requeue the analysisRun automatically after 15 seconds. We can check if the analysis run is completed by looking into the response for the completed field `.completed`: https://github.com/spinnaker/kayenta/blob/master/kayenta-core/src/main/java/com/netflix/kayenta/canary/CanaryExecutionStatusResponse.java#L40

Reply:
Added support to check for complete status and additional test cases.